rpcadaptors.go

```diff
@@ -324,3 +328,405 @@ func (b *rpcSyncMgr) ProcessTransaction(tx *dcrutil.Tx, allowOrphans bool,
+// implements the rpcserver.UtxoEntry interface.
+type rpcUtxoEntry struct {
+	utxoEntry *blockchain.UtxoEntry
+}
+
+// Ensure rpcUtxoEntry implements the rpcserver.UtxoEntry interface.
+var _ rpcserver.UtxoEntry = (*rpcUtxoEntry)(nil)
+
+// ToUtxoEntry returns the underlying UtxoEntry instance.
+func (u *rpcUtxoEntry) ToUtxoEntry() *blockchain.UtxoEntry {
+	return u.utxoEntry
+}
+
+// TransactionType returns the type of the transaction the utxo entry
+// represents.
+func (u *rpcUtxoEntry) TransactionType() stake.TxType {
+	return u.utxoEntry.TransactionType()
+}
+
+// IsOutputSpent returns whether or not the provided output index has been
+// spent based upon the current state of the unspent transaction output view
+// the entry was obtained from.
+//
+// Returns true if the output index references an output that does not exist
+// either due to it being invalid or because the output is not part of the view
+// due to previously being spent/pruned.
+func (u *rpcUtxoEntry) IsOutputSpent(outputIndex uint32) bool {
+	return u.utxoEntry.IsOutputSpent(outputIndex)
+}
+
+// BlockHeight returns the height of the block containing the transaction the
+// utxo entry represents.
+func (u *rpcUtxoEntry) BlockHeight() int64 {
+	return u.utxoEntry.BlockHeight()
+}
+
+// TxVersion returns the version of the transaction the utxo represents.
+func (u *rpcUtxoEntry) TxVersion() uint16 {
+	return u.utxoEntry.TxVersion()
+}
+
+// AmountByIndex returns the amount of the provided output index.
+//
+// Returns 0 if the output index references an output that does not exist
+// either due to it being invalid or because the output is not part of the view
+// due to previously being spent/pruned.
+func (u *rpcUtxoEntry) AmountByIndex(outputIndex uint32) int64 {
+	return u.utxoEntry.AmountByIndex(outputIndex)
+}
+
+// ScriptVersionByIndex returns the public key script for the provided output
+// index.
+//
+// Returns 0 if the output index references an output that does not exist
+// either due to it being invalid or because the output is not part of the view
+// due to previously being spent/pruned.
+func (u *rpcUtxoEntry) ScriptVersionByIndex(outputIndex uint32) uint16 {
+	return u.utxoEntry.ScriptVersionByIndex(outputIndex)
+}
+
+// PkScriptByIndex returns the public key script for the provided output index.
+//
+// Returns nil if the output index references an output that does not exist
+// either due to it being invalid or because the output is not part of the view
+// due to previously being spent/pruned.
+func (u *rpcUtxoEntry) PkScriptByIndex(outputIndex uint32) []byte {
+	return u.utxoEntry.PkScriptByIndex(outputIndex)
+}
+
+// IsCoinBase returns whether or not the transaction the utxo entry represents
+// is a coinbase.
+func (u *rpcUtxoEntry) IsCoinBase() bool {
+	return u.utxoEntry.IsCoinBase()
+}
+
+// rpcChain provides a chain for use with the RPC server and
+// implements the rpcserver.Chain interface.
+type rpcChain struct {
+	chain *blockchain.BlockChain
```

Comment:
The same thing is true here.  You can avoid hundreds of lines of code here by just embedding the type directly and defining only the cases that either don't exist or don't have matching signatures.  Namely, `ConvertUtxosToMinimalOutputs ` and `FetchUtxoEntry`.

e.g. The following will implement it properly:

```Go
// rpcChain provides a chain for use with the RPC server and
// implements the rpcserver.Chain interface.
type rpcChain struct {
	*blockchain.BlockChain
}

// Ensure rpcChain implements the rpcserver.Chain interface.
var _ rpcserver.Chain = (*rpcChain)(nil)

// ConvertUtxosToMinimalOutputs converts the contents of a UTX to a series of
// minimal outputs. It does this so that these can be passed to stake subpackage
// functions, where they will be evaluated for correctness.
func (c *rpcChain) ConvertUtxosToMinimalOutputs(entry rpcserver.UtxoEntry) []*stake.MinimalOutput {
	return blockchain.ConvertUtxosToMinimalOutputs(entry.ToUtxoEntry())
}

// FetchUtxoEntry loads and returns the unspent transaction output entry for the
// passed hash from the point of view of the end of the main chain.
//
// NOTE: Requesting a hash for which there is no data will NOT return an error.
// Instead both the entry and the error will be nil.  This is done to allow
// pruning of fully spent transactions.  In practice this means the caller must
// check if the returned entry is nil before invoking methods on it.
//
// This function is safe for concurrent access however the returned entry (if
// any) is NOT.
func (c *rpcChain) FetchUtxoEntry(txHash *chainhash.Hash) (rpcserver.UtxoEntry, error) {
	utxo, err := c.BlockChain.FetchUtxoEntry(txHash)
	return &rpcUtxoEntry{utxo}, err
}
```

Reply:
Makes sense, updated - I started this way to avoid exposing the underlying `blockchain.BlockChain` and its methods that are unrelated to what `rpcserver` requires. But since `rpcserver` is just using the `rpcserver.Chain` interface, and not the `rpcChain` adaptor directly, it doesn't matter.